horovod/common/operations.cc

```diff
@@ -873,81 +981,44 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
         }
         if (timeline.Initialized()) {
-          RECORD_EVENT(entries, after_memcpy_out_event, stream)
+          RECORD_EVENT("MEMCPY_OUT_FUSION_BUFFER",
+                       entries, after_memcpy_out_event, stream, false)
         }
       } else {
         auto e = first_entry;
+#if HOROVOD_GPU_ALLREDUCE == 'D'
+        // Copy input to output buffer because DDL runs in place
+        CUDA_CHECK(entries, "cudaMemcpyAsync",
+                   cudaMemcpyAsync((void*)e.output->data(), e.tensor->data(),
+                                   (size_t)e.tensor->shape().num_elements(),
+                                   cudaMemcpyDeviceToDevice, stream))
+        RECORD_EVENT("MEMCPY_IN_OUT_BUFFER",
+                     entries, after_memcpy_in_event, stream, false)
+        CUDA_CHECK(entries, "cudaEventSynchronize",
+                   cudaEventSynchronize(after_memcpy_in_event))
+
+        DDL_CHECK(entries, "ddl_allreduce",
+                  ddl_allreduce((void*)e.output->data(),
+                                (size_t)e.tensor->shape().num_elements(),
+                                GetDDLDataType(first_entry.tensor),
+                                DDL_OP_SUM))
+#else
         NCCL_CHECK(entries, "ncclAllReduce",
                    ncclAllReduce(e.tensor->data(), (void*)e.output->data(),
                                  (size_t)e.tensor->shape().num_elements(),
                                  GetNCCLDataType(first_entry.tensor), ncclSum,
                                  nccl_comm, stream))
+#endif
         if (timeline.Initialized()) {
-          RECORD_EVENT(entries, after_reduce_event, stream)
+          RECORD_EVENT("ALLREDUCE", entries, after_reduce_event, stream, false)
         }
       }
 
-      ACTIVITY_END_ALL(entries, timeline)
-      ACTIVITY_START_ALL(entries, timeline, "QUEUE")
-
       // Use completion marker via event because it's faster than
       // blocking cudaStreamSynchronize() in this thread.
       cudaEvent_t done_event;
-      RECORD_EVENT(entries, done_event, stream)
-
-      // TODO: use thread pool or single thread for callbacks
-      std::thread finalizer_thread([entries, first_entry, done_event,
-                                    queue_end_event, after_memcpy_in_event,
-                                    after_reduce_event, after_memcpy_out_event,
-                                    response, &timeline] {
-        CUDA_CHECK(entries, "cudaSetDevice", cudaSetDevice(first_entry.device))
-        if (queue_end_event != nullptr) {
-          CUDA_CHECK(entries, "cudaEventSynchronize",
-                     cudaEventSynchronize(queue_end_event))
-          // All the work scheduled on NCCL stream before this allreduce
-          // is done at this point, end queueing activity.
-          ACTIVITY_END_ALL(entries, timeline)
-          RELEASE_EVENT(entries, queue_end_event);
-        }
+      RECORD_EVENT("ALLREDUCE_DONE", entries, done_event, stream, true)
```

Comment:
Can we do this without changing the set of events we're tracking?

Reply:
This event is the done_event that was there before. You mean without adding an extra activity to the timeline? The bool set to true at the end of the call means it is the final event. We could not record the event as an activity in the finalizer thread. Would you prefer that?